core/src/test/scala/com.mediative.sparrow/DataFrameReaderTest.scala

```diff
@@ -173,51 +195,107 @@ class DataFrameReaderTest extends FreeSpec with BeforeAndAfterAll {
+        val rdd1 = rdd1Maybe.toOption.get
+        assertResult(0) { rdd0.subtract(rdd1).count }
+        assertResult(0) { rdd1.subtract(rdd0).count }
+      }
 
-      assert(df.toRDD[TestToRdd1].toOption.get.first == expected)
-    }
+      import TestCaseClasses._
 
-    "round-trip an object containing an Int and an optional Int from RDD to DataFrame back to RDD" - {
-      "when containing some value" in {
-        import TestCaseClasses.TestToRdd2
+      "Int, String" in {
+        testRDD_DataFrame_codec(TestToRdd1(1, "a"))
+      }
 
-        val expected = TestToRdd2(1, Option(1))
-        val df =
-          sqlContext.createDataFrame(sc.parallelize(List(expected)))
+      "Int, Option[Int]" - {
+        "when Some(Int)" in {
+          testRDD_DataFrame_codec(TestToRdd2(1, Option(1)))
+        }
 
-        assert(df.toRDD[TestToRdd2].toOption.get.first == expected)
+        "when None" in {
+          testRDD_DataFrame_codec(TestToRdd2(1, Option.empty))
+        }
       }
 
-      "when containing no value" in {
-        import TestCaseClasses.TestToRdd2
-
-        val expected = TestToRdd2(1, Option.empty)
-        val df =
-          sqlContext.createDataFrame(sc.parallelize(List(expected)))
+      "String, java.sql.Timestamp" in {
+        pendingUntilFixed {
+          // FIXME:
+          // rdd1Maybe.isSuccess was false Failure(NonEmptyList(The field 'timestampVal' isn't a LongType as expected, TimestampType received.)) (DataFrameReaderTest.scala:203)
+          testRDD_DataFrame_codec(
+            TestToRdd3("a", Timestamp.valueOf("2015-07-15 09:00:00"))
+          )
+        }
+      }
 
-        assert(df.toRDD[TestToRdd2].toOption.get.first == expected)
+      "Int, Double" in {
+        pendingUntilFixed {
+          // FIXME:
+          // "org.apache.spark.SparkException: Job aborted due to stage failure"
+          // Caused by: java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Integer
+          // at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:106)
+          testRDD_DataFrame_codec(TestToRdd4(1, 2.0))
+        }
       }
-    }
 
-    "round-trip an object containing a String and a Timestamp from RDD to DataFrame back to RDD" in {
-      import TestCaseClasses.TestToRdd3
+      "Int, Option[Double]" - {
+        "when Some(Double)" in {
+          testRDD_DataFrame_codec(TestToRdd5(1, Some(2.0)))
+        }
+        "when None" in {
+          testRDD_DataFrame_codec(TestToRdd5(1, None))
+        }
+      }
 
-      val expected =
-        TestToRdd3("a", Timestamp.valueOf("2015-07-15 09:00:00"))
+      "Int, Wrapped[String]" in {
+        pendingUntilFixed {
+          // FIXME:
+          // rdd1Maybe.isSuccess was false Failure(NonEmptyList(The field 'wrappedStringVal' isn't a StringType as expected, StructType(StructField(unwrap,StringType,true)) received.)) (DataFrameReaderTest.scala:207)
+          testRDD_DataFrame_codec(TestToRdd8(1, Wrap.Wrapped("foo")))
```

Comment:
@suhailshergill:

> could you please point me to the line in the current PR changeset where i am reading from Parquet?

I don't know if `createDataFrame` is actually part of Parquet or not, but that is the part I was referring to. My initial understanding was that Parquet was persisting case classes, so that would mean the conversion of a case class to a Spark `Row` would be done through Parquet, but it's quite possible that it's simply provided by Spark and Parquet has nothing to do with it, which would mean that what's crappy isn't Parquet but Spark's `createDataFrame`.

So what I'm saying is if you have a case class that you're transforming to a `Row` through `createDataFrame`, then you need to have a `RowConverter` that understand the same structure if you want to go back from the `Row` to the case class. Those issues are about defining a `RowConverter` schema that has a structure that doesn't match the one resulting from calling `createDataFrame` (what I unfortunately referred to as saved by Parquet earlier). If you define your [Field|Row]Converter to match the same structure, then everything will work. Most of those issues aren't in Sparrow, they are in `ReportSpark`. And that's where the fixes are as well. That is, until Sparrow can offer an alternative to `createDataFrame` and provide a better solution.

Otherwise, what change in Sparrow would you suggest to handle such use case? I don't see a solution to the `ReportSpark` problems being possible in Sparrow, that's why I don't understand why making it part of this PR. The fix that Sparrow can provide eventually is to offer an alternative to `createDataFrame`, but then it wouldn't make this test anymore relevant because it relies on `createDataFrame`.